src/etc/man/crabgo.1
'\" t
.TH "CRABGO" "1"
.nh
.ad l
.ss \n[.ss] 0
.SH "NAME"
crabgo \[em] The Rust package manager
.SH "SYNOPSIS"
\fBcrabgo\fR [\fIoptions\fR] \fIcommand\fR [\fIargs\fR]
.br
\fBcrabgo\fR [\fIoptions\fR] \fB\-\-version\fR
.br
\fBcrabgo\fR [\fIoptions\fR] \fB\-\-list\fR
.br
\fBcrabgo\fR [\fIoptions\fR] \fB\-\-help\fR
.br
\fBcrabgo\fR [\fIoptions\fR] \fB\-\-explain\fR \fIcode\fR
.SH "DESCRIPTION"
This program is a package manager and build tool for the Rust language,
available at <https://rust\-lang.org>\&.
.SH "COMMANDS"
.SS "Build Commands"
\fBcrabgo\-bench\fR(1)
.br
\ \ \ \ Execute benchmarks of a package.
.sp
\fBcrabgo\-build\fR(1)
.br
\ \ \ \ Compile a package.
.sp
\fBcrabgo\-check\fR(1)
.br
\ \ \ \ Check a local package and all of its dependencies for errors.
.sp
\fBcrabgo\-clean\fR(1)
.br
\ \ \ \ Remove artifacts that Crabgo has generated in the past.
.sp
\fBcrabgo\-doc\fR(1)
.br
\ \ \ \ Build a package\[cq]s documentation.
.sp
\fBcrabgo\-fetch\fR(1)
.br
\ \ \ \ Fetch dependencies of a package from the network.
.sp
\fBcrabgo\-fix\fR(1)
.br
\ \ \ \ Automatically fix lint warnings reported by rustc.
.sp
\fBcrabgo\-run\fR(1)
.br
\ \ \ \ Run a binary or example of the local package.
.sp
\fBcrabgo\-rustc\fR(1)
.br
\ \ \ \ Compile a package, and pass extra options to the compiler.
.sp
\fBcrabgo\-rustdoc\fR(1)
.br
\ \ \ \ Build a package\[cq]s documentation, using specified custom flags.
.sp
\fBcrabgo\-test\fR(1)
.br
\ \ \ \ Execute unit and integration tests of a package.
.SS "Manifest Commands"
\fBcrabgo\-generate\-lockfile\fR(1)
.br
\ \ \ \ Generate \fBCrabgo.lock\fR for a project.
.sp
\fBcrabgo\-locate\-project\fR(1)
.br
\ \ \ \ Print a JSON representation of a \fBCrabgo.toml\fR file\[cq]s location.
.sp
\fBcrabgo\-metadata\fR(1)
.br
\ \ \ \ Output the resolved dependencies of a package in machine\-readable format.
.sp
\fBcrabgo\-pkgid\fR(1)
.br
\ \ \ \ Print a fully qualified package specification.
.sp
\fBcrabgo\-tree\fR(1)
.br
\ \ \ \ Display a tree visualization of a dependency graph.
.sp
\fBcrabgo\-update\fR(1)
.br
\ \ \ \ Update dependencies as recorded in the local lock file.
.sp
\fBcrabgo\-vendor\fR(1)
.br
\ \ \ \ Vendor all dependencies locally.
.sp
\fBcrabgo\-verify\-project\fR(1)
.br
\ \ \ \ Check correctness of crate manifest.
.SS "Package Commands"
\fBcrabgo\-init\fR(1)
.br
\ \ \ \ Create a new Crabgo package in an existing directory.
.sp
\fBcrabgo\-install\fR(1)
.br
\ \ \ \ Build and install a Rust binary.
.sp
\fBcrabgo\-new\fR(1)
.br
\ \ \ \ Create a new Crabgo package.
.sp
\fBcrabgo\-search\fR(1)
.br
\ \ \ \ Search packages in crates.io.
.sp
\fBcrabgo\-uninstall\fR(1)
.br
\ \ \ \ Remove a Rust binary.
.SS "Publishing Commands"
\fBcrabgo\-login\fR(1)
.br
\ \ \ \ Save an API token from the registry locally.
.sp
\fBcrabgo\-logout\fR(1)
.br
\ \ \ \ Remove an API token from the registry locally.
.sp
\fBcrabgo\-owner\fR(1)
.br
\ \ \ \ Manage the owners of a crate on the registry.
.sp
\fBcrabgo\-package\fR(1)
.br
\ \ \ \ Assemble the local package into a distributable tarball.
.sp
\fBcrabgo\-publish\fR(1)
.br
\ \ \ \ Upload a package to the registry.
.sp
\fBcrabgo\-yank\fR(1)
.br
\ \ \ \ Remove a pushed crate from the index.
.SS "General Commands"
\fBcrabgo\-help\fR(1)
.br
\ \ \ \ Display help information about Crabgo.
.sp
\fBcrabgo\-version\fR(1)
.br
\ \ \ \ Show version information.
.SH "OPTIONS"
.SS "Special Options"
.sp
\fB\-V\fR, 
\fB\-\-version\fR
.RS 4
Print version info and exit. If used with \fB\-\-verbose\fR, prints extra
information.
.RE
.sp
\fB\-\-list\fR
.RS 4
List all installed Crabgo subcommands. If used with \fB\-\-verbose\fR, prints extra
information.
.RE
.sp
\fB\-\-explain\fR \fIcode\fR
.RS 4
Run \fBrustc \-\-explain CODE\fR which will print out a detailed explanation of an
error message (for example, \fBE0004\fR).
.RE
.SS "Display Options"
.sp
\fB\-v\fR, 
\fB\-\-verbose\fR
.RS 4
Use verbose output. May be specified twice for \[lq]very verbose\[rq] output which
includes extra output such as dependency warnings and build script output.
May also be specified with the \fBterm.verbose\fR
\fIconfig value\fR <https://doc.rust\-lang.org/crabgo/reference/config.html>\&.
.RE
.sp
\fB\-q\fR, 
\fB\-\-quiet\fR
.RS 4
Do not print crabgo log messages.
May also be specified with the \fBterm.quiet\fR
\fIconfig value\fR <https://doc.rust\-lang.org/crabgo/reference/config.html>\&.
.RE
.sp
\fB\-\-color\fR \fIwhen\fR
.RS 4
Control when colored output is used. Valid values:
.sp
.RS 4
\h'-04'\(bu\h'+02'\fBauto\fR (default): Automatically detect if color support is available on the
terminal.
.RE
.sp
.RS 4
\h'-04'\(bu\h'+02'\fBalways\fR: Always display colors.
.RE
.sp
.RS 4
\h'-04'\(bu\h'+02'\fBnever\fR: Never display colors.
.RE
.sp
May also be specified with the \fBterm.color\fR
\fIconfig value\fR <https://doc.rust\-lang.org/crabgo/reference/config.html>\&.
.RE
.SS "Manifest Options"
.sp
\fB\-\-frozen\fR, 
\fB\-\-locked\fR
.RS 4
Either of these flags requires that the \fBCrabgo.lock\fR file is
up\-to\-date. If the lock file is missing, or it needs to be updated, Crabgo will
exit with an error. The \fB\-\-frozen\fR flag also prevents Crabgo from
attempting to access the network to determine if it is out\-of\-date.
.sp
These may be used in environments where you want to assert that the
\fBCrabgo.lock\fR file is up\-to\-date (such as a CI build) or want to avoid network
access.
.RE
.sp
\fB\-\-offline\fR
.RS 4
Prevents Crabgo from accessing the network for any reason. Without this
flag, Crabgo will stop with an error if it needs to access the network and
the network is not available. With this flag, Crabgo will attempt to
proceed without the network if possible.
.sp
Beware that this may result in different dependency resolution than online
mode. Crabgo will restrict itself to crates that are downloaded locally, even
if there might be a newer version as indicated in the local copy of the index.
See the \fBcrabgo\-fetch\fR(1) command to download dependencies before going
offline.
.sp
May also be specified with the \fBnet.offline\fR \fIconfig value\fR <https://doc.rust\-lang.org/crabgo/reference/config.html>\&.
.RE
.SS "Common Options"
.sp
\fB+\fR\fItoolchain\fR
.RS 4
If Crabgo has been installed with rustup, and the first argument to \fBcrabgo\fR
begins with \fB+\fR, it will be interpreted as a rustup toolchain name (such
as \fB+stable\fR or \fB+nightly\fR).
See the \fIrustup documentation\fR <https://rust\-lang.github.io/rustup/overrides.html>
for more information about how toolchain overrides work.
.RE
.sp
\fB\-\-config\fR \fIKEY=VALUE\fR or \fIPATH\fR
.RS 4
Overrides a Crabgo configuration value. The argument should be in TOML syntax of \fBKEY=VALUE\fR,
or provided as a path to an extra configuration file. This flag may be specified multiple times.
See the \fIcommand\-line overrides section\fR <https://doc.rust\-lang.org/crabgo/reference/config.html#command\-line\-overrides> for more information.
.RE
.sp
\fB\-C\fR \fIPATH\fR
.RS 4
Changes the current working directory before executing any specified operations. This affects
things like where crabgo looks by default for the project manifest (\fBCrabgo.toml\fR), as well as
the directories searched for discovering \fB\&.crabgo/config.toml\fR, for example. This option must
appear before the command name, for example \fBcrabgo \-C path/to/my\-project build\fR\&.
.sp
This option is only available on the \fInightly
channel\fR <https://doc.rust\-lang.org/book/appendix\-07\-nightly\-rust.html> and
requires the \fB\-Z unstable\-options\fR flag to enable (see
\fI#10098\fR <https://github.com/rust\-lang/crabgo/issues/10098>).
.RE
.sp
\fB\-h\fR, 
\fB\-\-help\fR
.RS 4
Prints help information.
.RE
.sp
\fB\-Z\fR \fIflag\fR
.RS 4
Unstable (nightly\-only) flags to Crabgo. Run \fBcrabgo \-Z help\fR for details.
.RE
.SH "ENVIRONMENT"
See \fIthe reference\fR <https://doc.rust\-lang.org/crabgo/reference/environment\-variables.html> for
details on environment variables that Crabgo reads.
.SH "EXIT STATUS"
.sp
.RS 4
\h'-04'\(bu\h'+02'\fB0\fR: Crabgo succeeded.
.RE
.sp
.RS 4
\h'-04'\(bu\h'+02'\fB101\fR: Crabgo failed to complete.
.RE
.SH "FILES"
\fB~/.crabgo/\fR
.br
\ \ \ \ Default location for Crabgo\[cq]s \[lq]home\[rq] directory where it
stores various files. The location can be changed with the \fBCRABGO_HOME\fR
environment variable.
.sp
\fB$CRABGO_HOME/bin/\fR
.br
\ \ \ \ Binaries installed by \fBcrabgo\-install\fR(1) will be located here. If using
\fIrustup\fR <https://rust\-lang.github.io/rustup/>, executables distributed with Rust are also located here.
.sp
\fB$CRABGO_HOME/config.toml\fR
.br
\ \ \ \ The global configuration file. See \fIthe reference\fR <https://doc.rust\-lang.org/crabgo/reference/config.html>
for more information about configuration files.
.sp
\fB\&.crabgo/config.toml\fR
.br
\ \ \ \ Crabgo automatically searches for a file named \fB\&.crabgo/config.toml\fR in the
current directory, and all parent directories. These configuration files
will be merged with the global configuration file.
.sp
\fB$CRABGO_HOME/credentials.toml\fR
.br
\ \ \ \ Private authentication information for logging in to a registry.
.sp
\fB$CRABGO_HOME/registry/\fR
.br
\ \ \ \ This directory contains cached downloads of the registry index and any
downloaded dependencies.
.sp
\fB$CRABGO_HOME/git/\fR
.br
\ \ \ \ This directory contains cached downloads of git dependencies.
.sp
Please note that the internal structure of the \fB$CRABGO_HOME\fR directory is not
stable yet and may be subject to change.
.SH "EXAMPLES"
.sp
.RS 4
\h'-04' 1.\h'+01'Build a local package and all of its dependencies:
.sp
.RS 4
.nf
crabgo build
.fi
.RE
.RE
.sp
.RS 4
\h'-04' 2.\h'+01'Build a package with optimizations:
.sp
.RS 4
.nf
crabgo build \-\-release
.fi
.RE
.RE
.sp
.RS 4
\h'-04' 3.\h'+01'Run tests for a cross\-compiled target:
.sp
.RS 4
.nf
crabgo test \-\-target i686\-unknown\-linux\-gnu
.fi
.RE
.RE
.sp
.RS 4
\h'-04' 4.\h'+01'Create a new package that builds an executable:
.sp
.RS 4
.nf
crabgo new foobar
.fi
.RE
.RE
.sp
.RS 4
\h'-04' 5.\h'+01'Create a package in the current directory:
.sp
.RS 4
.nf
mkdir foo && cd foo
crabgo init .
.fi
.RE
.RE
.sp
.RS 4
\h'-04' 6.\h'+01'Learn about a command\[cq]s options and usage:
.sp
.RS 4
.nf
crabgo help clean
.fi
.RE
.RE
.SH "BUGS"
See <https://github.com/rust\-lang/crabgo/issues> for issues.
.SH "SEE ALSO"
\fBrustc\fR(1), \fBrustdoc\fR(1)


CHANGELOG.md
# Changelog

## Crabgo 1.71 (2023-07-13)
[84b7041f...HEAD](https://github.com/rust-lang/crabgo/compare/84b7041f...HEAD)

### Added

### Changed

- Better error message when getting an empty dependency table in Crabgo.toml.
  [#11997](https://github.com/rust-lang/crabgo/pull/11997)
- Use restricted Damerau-Levenshtein algorithm to provide typo suggestions.
  [#11963](https://github.com/rust-lang/crabgo/pull/11963)

### Fixed

- Corrected the bug report URL for `crabgo clippy --fix`.
  [#11882](https://github.com/rust-lang/crabgo/pull/11882)

### Nightly only

- Recompile on profile rustflags changes.
  [#11981](https://github.com/rust-lang/crabgo/pull/11981)

### Documentation

### Internal

- Crabgo is now a Crabgo workspace. We dogfood ourselves finally.
  [#11851](https://github.com/rust-lang/crabgo/pull/11851)
  [#11994](https://github.com/rust-lang/crabgo/pull/11994)
  [#11996](https://github.com/rust-lang/crabgo/pull/11996)
- Allow win/mac credential managers to build on all platforms.
  [#11993](https://github.com/rust-lang/crabgo/pull/11993)
- Use `openssl` only on non-Windows platforms.
  [#11979](https://github.com/rust-lang/crabgo/pull/11979)
- A new, straightforward issue labels system for Crabgo contributors.
  [docs](https://doc.crates.io/contrib/issues.html)
  [#11995](https://github.com/rust-lang/crabgo/pull/11995)
  [#12002](https://github.com/rust-lang/crabgo/pull/12002)
  [#12003](https://github.com/rust-lang/crabgo/pull/12003)

## Crabgo 1.70 (2023-06-01)
[9880b408...rust-1.70.0](https://github.com/rust-lang/crabgo/compare/9880b408...rust-1.70.0)

### Added

- 🎉 Added `crabgo logout` command for removing an API token from the registry locally.
  [docs](https://doc.rust-lang.org/nightly/crabgo/commands/crabgo-logout.html)
  [#11919](https://github.com/rust-lang/crabgo/pull/11919)
  [#11950](https://github.com/rust-lang/crabgo/pull/11950)
- Added `--ignore-rust-version` flag to `crabgo install`.
  [#11859](https://github.com/rust-lang/crabgo/pull/11859)
- The `CRABGO_PKG_README` environment variable is now set to the path to the
  README file when compiling a crate.
  [#11645](https://github.com/rust-lang/crabgo/pull/11645)
- Crabgo now displays richer information of Crabgo target failed to compile.
  [#11636](https://github.com/rust-lang/crabgo/pull/11636)

### Changed

- 🎉 The `sparse` protocol is now the default protocol for crates.io!
  ([RFC 2789](https://github.com/rust-lang/rfcs/blob/master/text/2789-sparse-index.md))
  ([docs](https://doc.rust-lang.org/nightly/crabgo/reference/registries.html#registry-protocols))
  [#11791](https://github.com/rust-lang/crabgo/pull/11791)
  [#11783](https://github.com/rust-lang/crabgo/pull/11783)
- ❗ `crabgo login` and `crabgo logout` now uses the registry specified in
  `registry.default`. This was an unintentional regression.
  [#11949](https://github.com/rust-lang/crabgo/pull/11949)
- `crabgo update` accurately shows `Downgrading` status when downgrading dependencies.
  [#11839](https://github.com/rust-lang/crabgo/pull/11839)
- Added more information to HTTP errors to help with debugging.
  [#11878](https://github.com/rust-lang/crabgo/pull/11878)
- Added delays to network retries in Crabgo.
  [#11881](https://github.com/rust-lang/crabgo/pull/11881)
- Refined `crabgo puslish` message when waiting for a publish complete.
  [#11713](https://github.com/rust-lang/crabgo/pull/11713)
- Better error message when `crabgo install` from a git repository but found
  multiple packages.
  [#11835](https://github.com/rust-lang/crabgo/pull/11835)

### Fixed

- Removed duplicates of possible values in `--charset` option of `crabgo tree`.
  [#11785](https://github.com/rust-lang/crabgo/pull/11785)
- Fixed `CRABGO_CFG_` vars for configs defined both with and without value.
  [#11790](https://github.com/rust-lang/crabgo/pull/11790)
- Broke endless loop on cyclic features in added dependency in `crabgo add`.
  [#11805](https://github.com/rust-lang/crabgo/pull/11805)
- Don't panic when [`patch`] involved in dependency resolution results in a conflict.
  [#11770](https://github.com/rust-lang/crabgo/pull/11770)
- Fixed credential token format validation.
  [#11951](https://github.com/rust-lang/crabgo/pull/11951)
- Added the missing token format validation on publish.
  [#11952](https://github.com/rust-lang/crabgo/pull/11952)
- Fixed case mismatches when looking up env vars in the Config snapshot.
  [#11824](https://github.com/rust-lang/crabgo/pull/11824)
- `crabgo new` generates the correct `.hgignore` aligning semantics with other
  VCS ignore files.
  [#11855](https://github.com/rust-lang/crabgo/pull/11855)

### Nightly only

- Added `-Zdirect-minimal-versions`. This behaves like `-Zminimal-versions` but
  only for direct dependencies.
  ([docs](https://doc.rust-lang.org/nightly/crabgo/reference/unstable.html#direct-minimal-versions))
  [#11688](https://github.com/rust-lang/crabgo/pull/11688)
- Added `-Zgitoxide` which switches all `git fetch` operation in Crabgo to
  use `gitoxide` crate. This is still an MVP but could improve the performance
  up to 2 times.
  ([docs](https://doc.rust-lang.org/nightly/crabgo/reference/unstable.html##gitoxide))
  [#11448](https://github.com/rust-lang/crabgo/pull/11448)
  [#11800](https://github.com/rust-lang/crabgo/pull/11800)
  [#11822](https://github.com/rust-lang/crabgo/pull/11822)
  [#11830](https://github.com/rust-lang/crabgo/pull/11830)
- Removed `-Zjobserver-per-rustc`. Its rustc counterpart never got landed.
  [#11764](https://github.com/rust-lang/crabgo/pull/11764)

### Documentation

- Cleaned-up unstable documentation.
  [#11793](https://github.com/rust-lang/crabgo/pull/11793)
- Enhanced the documentation of timing report with graphs.
  [#11798](https://github.com/rust-lang/crabgo/pull/11798)
- Clarified requirements about the state of the registry index after publish.
  [#11926](https://github.com/rust-lang/crabgo/pull/11926)
- Clarified docs on `-C` that it appears before the command.

src/etc/crabgo.bashcomp.sh
# Required for bash versions < 4.1
# Default bash version is 3.2 on latest macOS. See #6874
shopt -s extglob

command -v crabgo >/dev/null 2>&1 &&
_crabgo()
{
	local cur prev words cword
	_get_comp_words_by_ref cur prev words cword

	COMPREPLY=()

	# Skip past - and + options to find the command.
	local nwords=${#words[@]}
	local cmd_i cmd dd_i
	for (( cmd_i=1; cmd_i<$nwords; cmd_i++ ));
	do
		if [[ ! "${words[$cmd_i]}" =~ ^[+-] ]]; then
			cmd="${words[$cmd_i]}"
			break
		fi
	done
	# Find the location of the -- separator.
	for (( dd_i=1; dd_i<$nwords-1; dd_i++ ));
	do
		if [[ "${words[$dd_i]}" = "--" ]]; then
			break
		fi
	done

	local vcs='git hg none pijul fossil'
	local color='auto always never'
	local msg_format='human json short'

	local opt_help='-h --help'
	local opt_verbose='-v --verbose'
	local opt_quiet='-q --quiet'
	local opt_color='--color'
	local opt_common="$opt_help $opt_verbose $opt_quiet $opt_color"
	local opt_pkg_spec='-p --package --all --exclude --workspace'
	local opt_pkg='-p --package'
	local opt_feat='-F --features --all-features --no-default-features'
	local opt_mani='--manifest-path'
	local opt_parallel='-j --jobs --keep-going'
	local opt_force='-f --force'
	local opt_sync='-s --sync'
	local opt_lock='--frozen --locked --offline'
	local opt_targets="--lib --bin --bins --example --examples --test --tests --bench --benches --all-targets"

	local opt___nocmd="$opt_common -V --version --list --explain"
	local opt__add="$opt_common -p --package --features --default-features --no-default-features $opt_mani --optional --no-optional --rename --dry-run --path --git --branch --tag --rev --registry --dev --build --target"
	local opt__bench="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets --message-format --target --no-run --no-fail-fast --target-dir --ignore-rust-version"
	local opt__build="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets --message-format --target --release --profile --target-dir --ignore-rust-version"
	local opt__b="$opt__build"
	local opt__check="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets --message-format --target --release --profile --target-dir --ignore-rust-version"
	local opt__c="$opt__check"
	local opt__clean="$opt_common $opt_pkg $opt_mani $opt_lock --target --release --doc --target-dir --profile"
	local opt__clippy="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets --message-format --target --release --profile --target-dir --no-deps --fix"
	local opt__doc="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock $opt_parallel --message-format --bin --bins --lib --target --open --no-deps --release --document-private-items --target-dir --profile --ignore-rust-version"
	local opt__d="$opt__doc"
	local opt__fetch="$opt_common $opt_mani $opt_lock --target"
	local opt__fix="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_parallel $opt_targets $opt_lock --release --target --message-format --broken-code --edition --edition-idioms --allow-no-vcs --allow-dirty --allow-staged --profile --target-dir --ignore-rust-version"
	local opt__generate_lockfile="$opt_common $opt_mani $opt_lock"
	local opt__help="$opt_help"
	local opt__init="$opt_common $opt_lock --bin --lib --name --vcs --edition --registry"
	local opt__install="$opt_common $opt_feat $opt_parallel $opt_lock $opt_force --bin --bins --branch --debug --example --examples --git --list --path --rev --root --tag --version --registry --target --profile --no-track --ignore-rust-version"
	local opt__locate_project="$opt_common $opt_mani $opt_lock --message-format --workspace"
	local opt__login="$opt_common $opt_lock --registry"
	local opt__metadata="$opt_common $opt_feat $opt_mani $opt_lock --format-version=1 --no-deps --filter-platform"
	local opt__new="$opt_common $opt_lock --vcs --bin --lib --name --edition --registry"
	local opt__owner="$opt_common $opt_lock -a --add -r --remove -l --list --index --token --registry"
	local opt__package="$opt_common $opt_mani $opt_feat $opt_lock $opt_parallel --allow-dirty -l --list --no-verify --no-metadata --target --target-dir"
	local opt__pkgid="$opt_common $opt_mani $opt_lock $opt_pkg"
	local opt__publish="$opt_common $opt_mani $opt_feat $opt_lock $opt_parallel --allow-dirty --dry-run --token --no-verify --index --registry --target --target-dir"
	local opt__read_manifest="$opt_help $opt_quiet $opt_verbose $opt_mani $opt_color $opt_lock --no-deps"
	local opt__remove="$opt_common $opt_pkg $opt_lock $opt_mani --dry-run --dev --build --target"
	local opt__rm="$opt__remove"
	local opt__report="$opt_help $opt_verbose $opt_color future-incompat future-incompatibilities"
	local opt__report__future_incompat="$opt_help $opt_verbose $opt_color $opt_pkg --id"
	local opt__run="$opt_common $opt_pkg $opt_feat $opt_mani $opt_lock $opt_parallel --message-format --target --bin --example --release --target-dir --profile --ignore-rust-version"
	local opt__r="$opt__run"
	local opt__rustc="$opt_common $opt_pkg $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets -L --crate-type --extern --message-format --profile --target --release --target-dir --ignore-rust-version"
	local opt__rustdoc="$opt_common $opt_pkg $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets --message-format --target --release --open --target-dir --profile --ignore-rust-version"
	local opt__search="$opt_common $opt_lock --limit --index --registry"
	local opt__test="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock $opt_parallel $opt_targets --message-format --doc --target --no-run --release --no-fail-fast --target-dir --profile --ignore-rust-version"
	local opt__t="$opt__test"
	local opt__tree="$opt_common $opt_pkg_spec $opt_feat $opt_mani $opt_lock --target -i --invert --prefix --no-dedupe --duplicates -d --charset -f --format -e --edges"
	local opt__uninstall="$opt_common $opt_lock $opt_pkg --bin --root"
	local opt__update="$opt_common $opt_mani $opt_lock $opt_pkg --aggressive --precise --dry-run"
	local opt__vendor="$opt_common $opt_mani $opt_lock $opt_sync --no-delete --respect-source-config --versioned-dirs"
	local opt__verify_project="$opt_common $opt_mani $opt_lock"
	local opt__version="$opt_common $opt_lock"
	local opt__yank="$opt_common $opt_lock --version --undo --index --token --registry"
	local opt__libtest="--help --include-ignored --ignored --test --bench --list --logfile --nocapture --test-threads --skip -q --quiet --exact --color --format"

	if [[ $cword -gt $dd_i ]]; then
		# Completion after -- separator.
		if [[ "${cmd}" = @(test|bench) ]]; then
			COMPREPLY=( $( compgen -W "${opt__libtest}" -- "$cur" ) )
		else
			# Fallback to filename completion, useful with `crabgo run`.
			_filedir
		fi
	elif [[ $cword -le $cmd_i ]]; then
		# Completion before or at the command.
		if [[ "$cur" == -* ]]; then
			COMPREPLY=( $( compgen -W "${opt___nocmd}" -- "$cur" ) )
		elif [[ "$cur" == +* ]]; then
			COMPREPLY=( $( compgen -W "$(_toolchains)" -- "$cur" ) )
		else
			_ensure_crabgo_commands_cache_filled
			COMPREPLY=( $( compgen -W "$__crabgo_commands_cache" -- "$cur" ) )
		fi
	else
		case "${prev}" in
			--vcs)
				COMPREPLY=( $( compgen -W "$vcs" -- "$cur" ) )
				;;
			--color)
				COMPREPLY=( $( compgen -W "$color" -- "$cur" ) )
				;;
			--message-format)
				COMPREPLY=( $( compgen -W "$msg_format" -- "$cur" ) )
				;;
			--manifest-path)
				_filedir toml
				;;
			--bin)
				COMPREPLY=( $( compgen -W "$(_bin_names)" -- "$cur" ) )
				;;
			--test)
				COMPREPLY=( $( compgen -W "$(_test_names)" -- "$cur" ) )
				;;
			--bench)
				COMPREPLY=( $( compgen -W "$(_benchmark_names)" -- "$cur" ) )
				;;
			--example)
				COMPREPLY=( $( compgen -W "$(_get_examples)" -- "$cur" ) )
				;;
			--target)
				COMPREPLY=( $( compgen -W "$(_get_targets)" -- "$cur" ) )
				;;
			--target-dir|--path)
				_filedir -d
				;;
			help)
				_ensure_crabgo_commands_cache_filled
				COMPREPLY=( $( compgen -W "$__crabgo_commands_cache" -- "$cur" ) )
				;;
			*)
				if [[ "$cmd" == "report" && "$prev" == future-incompat* ]]; then
					local opt_var=opt__${cmd//-/_}__${prev//-/_}
				else
					local opt_var=opt__${cmd//-/_}
				fi
				if [[ -z "${!opt_var}" ]]; then
					# Fallback to filename completion.
					_filedir
				else
					COMPREPLY=( $( compgen -W "${!opt_var}" -- "$cur" ) )
				fi
				;;
		esac
	fi

	# compopt does not work in bash version 3

	return 0
} &&
complete -F _crabgo crabgo

__crabgo_commands_cache=
_ensure_crabgo_commands_cache_filled(){
	if [[ -z $__crabgo_commands_cache ]]; then
		__crabgo_commands_cache="$(crabgo --list 2>/dev/null | awk 'NR>1 {print $1}')"
	fi
}

_locate_manifest(){
	crabgo locate-project --message-format plain 2>/dev/null
}

# Extracts the values of "name" from the array given in $1 and shows them as
# command line options for completion
_get_names_from_array()
{
	local manifest=$(_locate_manifest)
	if [[ -z $manifest ]]; then
		return 0
	fi

	local last_line
	local -a names
	local in_block=false
	local block_name=$1
	while read line
	do
		if [[ $last_line == "[[$block_name]]" ]]; then
			in_block=true
		else
			if [[ $last_line =~ .*\[\[.* ]]; then
				in_block=false
			fi
		fi

		if [[ $in_block == true ]]; then
			if [[ $line =~ .*name.*\= ]]; then
				line=${line##*=}
				line=${line%%\"}
				line=${line##*\"}
				names+=($line)
			fi
		fi

		last_line=$line
	done < $manifest
	echo "${names[@]}"
}

#Gets the bin names from the manifest file
_bin_names()
{
	_get_names_from_array "bin"
}

#Gets the test names from the manifest file
_test_names()
{
	_get_names_from_array "test"
}

#Gets the bench names from the manifest file
_benchmark_names()
{
	_get_names_from_array "bench"
}

_get_examples(){
	local manifest=$(_locate_manifest)
	[ -z "$manifest" ] && return 0

	local files=("${manifest%/*}"/examples/*.rs)
	local names=("${files[@]##*/}")
	local names=("${names[@]%.*}")
	# "*" means no examples found
	if [[ "${names[@]}" != "*" ]]; then
		echo "${names[@]}"
	fi
}

_get_targets(){
	local result=()
	local targets=$(rustup target list)
	while read line
	do
		if [[ "$line" =~ default|installed ]]; then
			result+=("${line%% *}")
		fi
	done <<< "$targets"
	echo "${result[@]}"
}

_toolchains(){
	local result=()
	local toolchains=$(rustup toolchain list)
	local channels="nightly|beta|stable|[0-9]\.[0-9]{1,2}\.[0-9]"
	local date="[0-9]{4}-[0-9]{2}-[0-9]{2}"
	while read line
	do
		# Strip " (default)"
		line=${line%% *}
		if [[ "$line" =~ ^($channels)(-($date))?(-.*) ]]; then
			if [[ -z ${BASH_REMATCH[3]} ]]; then
				result+=("+${BASH_REMATCH[1]}")
			else
				# channel-date
				result+=("+${BASH_REMATCH[1]}-${BASH_REMATCH[3]}")
			fi
			result+=("+$line")
		else
			result+=("+$line")
		fi
	done <<< "$toolchains"
	echo "${result[@]}"
}

# vim:ft=sh
